### ISSUE TITLE
Add http.client.HTTPConnection instrumentation

### ISSUE THREAD

**Issue by goodoldneon** [elastic/apm-agent-python#1714]
**Is your feature request related to a problem? Please describe.**

Duo's client isn't instrumented because they use `http.client.HTTPConnection`.

**Describe the solution you'd like**

Add instrumentation for `http.client.HTTPConnection`.

**Describe alternatives you've considered**

Manually creating spans around Duo client calls, which is cumbersome and easy to forget.

**Additional context**

The `urllib3` instrumentation _almost_ works for `http.client.HTTPConnection`. It successfully instrumented Duo's client when I made these changes:

```diff
diff --git a/elasticapm/instrumentation/packages/urllib3.py b/elasticapm/instrumentation/packages/urllib3.py
index cce2c17c..65eba2b4 100644
--- a/elasticapm/instrumentation/packages/urllib3.py
+++ b/elasticapm/instrumentation/packages/urllib3.py
@@ -85,6 +85,7 @@ class Urllib3Instrumentation(AbstractInstrumentedModule):
         # packages that vendor or vendored urllib3 in the past
         ("requests.packages.urllib3.connectionpool", "HTTPConnectionPool.urlopen"),
         ("botocore.vendored.requests.packages.urllib3.connectionpool", "HTTPConnectionPool.urlopen"),
+        ("http.client", "HTTPConnection.request"),
     ]
 
     def call(self, module, method, wrapped, instance, args, kwargs):
@@ -95,8 +96,8 @@ class Urllib3Instrumentation(AbstractInstrumentedModule):
 
         host = instance.host
 
-        if instance.port != default_ports.get(instance.scheme):
-            host += ":" + str(instance.port)
+        # if instance.port != default_ports.get(instance.scheme):
+        #     host += ":" + str(instance.port)
 
         if "url" in kwargs:
             url = kwargs["url"]
@@ -105,8 +106,8 @@ class Urllib3Instrumentation(AbstractInstrumentedModule):
 
         signature = method.upper() + " " + host
 
-        if url.startswith("/"):
-            url = "%s://%s%s" % (instance.scheme, host, url)
+        # if url.startswith("/"):
+        #     url = "%s://%s%s" % (instance.scheme, host, url)
 
         transaction = execution_context.get_transaction()
```

Note how the `instance.scheme` stuff is commented out. There isn't a `scheme` attribute on `http.client.HTTPConnection`.

**Comment by basepi** [comment]
Hey @goodoldneon! We should definitely add an instrumentation here. Do you want to open a PR since you've done the initial legwork? We can work through issues and add some testing there.

Or, if you'd prefer, one of us can open the PR.

**Comment by goodoldneon** [comment]
@basepi I'd appreciate y'all creating the PR. We have a newborn so I'm my time is limited 😅

**Comment by beniwohli** [comment]
Hey @goodoldneon 

I took a stab at implementing this. Unfortunately, I ran into a bit of a roadblock. The core of the problem is that to do a complete request/response, the user of `http.client` needs to call two methods: `HTTPConnection.request()` and `HTTPConnection.getresponse()`. Our instrumentation framework isn't really set up for such scenarios, it is made to instrument and measure a single call.

If we just instrument `HTTPConnection.request()`, we only really measure how long it takes to send the HTTP headers to the server, and not how long it takes for the server to respond. We also miss out on information on the response (crucially, the response status).

Similarly, only instrumenting `HTTPConnection.getresponse()` means we don't get any request metadata (URL, headers, etc). Also, distributed tracing wouldn't work, as we need to modify the HTTP headers for that.

With enough bending, we could push our instrumentation API into kinda, sorta work (see #1788). It has several drawbacks, though:
 * we store some of our own state on `HTTPConnection`
 * we run into issues with calls to `request` and `getresponse` aren't balanced, e.g. if there is an exception between the two calls

If this was about any other library, I could be convinced that it's worth it to break the rules a little bit to make it work. But `http.client` is used by `urllib` and `urllib3` (and in turn, `requests`), so meddling with it is pretty high risk.

I'm afraid we'll have to close this for now.
